src/reflect/scala/reflect/internal/util/ScalaClassLoader.scala

```diff
@@ -33,88 +33,115 @@ trait HasClassPath {
  *  of java reflection.
  */
 trait ScalaClassLoader extends JClassLoader {
+  private def wrap = new ScalaClassLoader.RichJURLClassLoader(this)
   /** Executing an action with this classloader as context classloader */
-  def asContext[T](action: => T): T = {
-    import ScalaClassLoader.setContext
-
-    val saved = Thread.currentThread.getContextClassLoader
-    try { setContext(this) ; action }
-    finally setContext(saved)
-  }
+  def asContext[T](action: => T): T = wrap.asContext(action)
```

Comment:
Looks needless to hard-forward these.  The implicit conversion of `java.net.URLClassLoader`s in ScalaClassLoader's implicit scope seem to be enough for callsites to get these methods:

```
scala> import java.lang.{ ClassLoader => CL }, java.net.{ URLClassLoader => UCL }
import java.lang.{ClassLoader=>CL}
import java.net.{URLClassLoader=>UCL}

scala> trait SCL extends CL; object SCL { implicit class RichUCL(private val self: CL) extends AnyVal { def hi = println("hi") } }
trait SCL
object SCL

scala> val loader = new CL(getClass.getClassLoader) with SCL
val loader: ClassLoader with SCL = $anon$1@4177d422

scala> loader.hi
hi
```

This is strictly internal utility, right?  No bincompat worries?

If we must keep these, let's at least avoid forking the scaladoc (I think keep this one here, as it's more likely to be seen in an IDE compared to enrichment ones?)

Also, I'd define the AnyVal class here, top-level, so the history stays with the implementation, and hand define the implicit conversion on the companion, but now we're talking shed colours.

Reply:
I opted to forward them all to be binary compatibile with any unknown third party code using this API.